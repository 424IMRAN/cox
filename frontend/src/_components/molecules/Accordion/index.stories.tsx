import { Story, Meta } from '@storybook/react'
import { ThemeProvider } from '@material-ui/core/styles'
import AddIcon from '@material-ui/icons/Add'
import RemoveIcon from '@material-ui/icons/Remove'
import Accordion from '.'
import baseTheme from '../../../core-utils/theme'
import Typography from '../../atoms/Typography'

export default {
  title: 'molecules/Accordion',
  component: Accordion,
  argTypes: {
    expanded: {
      options: [false, true],
    },
    showBoxShadow: {
      options: [false, true],
    },
  },
} as Meta

const Template: Story = ({ ...args }) => (
  <ThemeProvider theme={baseTheme}>
    <Accordion
      header="Saved Filters"
      expandIcon={null}
      shrinkIcon={null}
      {...args}
    >
      <Typography variant="caption2">
        Lorem ipsum dolor sit amet consectetur adipisicing elit. Quas adipisci
        nihil sunt ex perferendis vel reiciendis aliquam dignissimos, hic ipsam
        harum quos itaque, omnis temporibus sequi odio magni consequatur
        dolores!
      </Typography>
    </Accordion>
  </ThemeProvider>
)

export const CustomIconAccordion = Template.bind({})
CustomIconAccordion.args = {
  expanded: true,
  header: 'Saved Filters',
  showBoxShadow: false,
  expandIcon: <AddIcon />,
  shrinkIcon: <RemoveIcon />,
}

export const BasicAccordion = Template.bind({})
BasicAccordion.args = {
  expanded: true,
  header: 'Saved Filters',
  showBoxShadow: false,
  showActive: false,
}

export const BoxShadowAccordion = Template.bind({})
BoxShadowAccordion.args = {
  expanded: true,
  header: 'Saved Filters',
  showBoxShadow: true,
}
