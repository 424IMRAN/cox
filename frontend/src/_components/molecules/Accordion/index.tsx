import React from 'react'
import { makeStyles, Theme, createStyles } from '@material-ui/core/styles'
import MuiAccordion from '@material-ui/core/Accordion'
import AccordionDetails from '@material-ui/core/AccordionDetails'
import AccordionSummary from '@material-ui/core/AccordionSummary'
import ExpandMoreIcon from '@material-ui/icons/ExpandMore'
import Typography from '../../atoms/Typography'
import { COLORS } from '../../../core-utils/theme'

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      width: '100%',
    },
    heading: {
      color: COLORS.TEXT_500,
    },
    headingActive: {
      color: COLORS.TEXT_100,
    },
    secondaryHeading: {
      fontSize: theme.typography.pxToRem(15),
      color: theme.palette.text.secondary,
    },
    noBoxShadow: {
      boxShadow: 'none',
    },
  })
)

type AccordionProps = {
  expanded?: boolean
  setExpanded?: (arg0: boolean) => void
  header: string
  children: any
  expandIcon?: React.ReactNode
  shrinkIcon?: React.ReactNode
  showBoxShadow?: boolean
  showActive?: boolean
}

const Accordion = ({
  expanded,
  setExpanded,
  header,
  children,
  expandIcon,
  shrinkIcon,
  showBoxShadow,
  showActive,
}: AccordionProps) => {
  const classes = useStyles()

  const handleChange = (
    event: React.ChangeEvent<Record<string, unknown>>,
    isExpanded: boolean
  ) => {
    setExpanded?.(!isExpanded)
  }

  const getActiveIcon = (icon: any) =>
    icon
      ? React.cloneElement(icon, {
          className: showActive ? classes.headingActive : classes.heading,
        })
      : null

  return (
    <div className={classes.root}>
      <MuiAccordion
        expanded={expanded}
        onChange={handleChange}
        className={!showBoxShadow ? classes.noBoxShadow : undefined}
      >
        <AccordionSummary
          expandIcon={
            ((expandIcon || shrinkIcon) &&
              (!expanded
                ? expandIcon ?? shrinkIcon
                : getActiveIcon(shrinkIcon ?? expandIcon))) ?? (
              <ExpandMoreIcon
                className={
                  expanded && showActive
                    ? classes.headingActive
                    : classes.heading
                }
              />
            )
          }
        >
          <Typography
            variant="subtitle2"
            className={
              expanded && showActive ? classes.headingActive : classes.heading
            }
          >
            {header}
          </Typography>
        </AccordionSummary>
        <AccordionDetails>{children}</AccordionDetails>
      </MuiAccordion>
    </div>
  )
}

Accordion.propTypes = {}

Accordion.defaultProps = {
  expanded: false,
  setExpanded: null,
  expandIcon: null,
  shrinkIcon: null,
  showBoxShadow: false,
  showActive: true,
}

export default Accordion
