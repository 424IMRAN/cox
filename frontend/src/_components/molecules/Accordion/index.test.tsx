import React from 'react'
import MuiAccordion from '@material-ui/core/Accordion'
import { mount, shallow } from 'enzyme'
import Accordion from '.'

describe('Accordion  Component', () => {
  it('Renders correctly', () => {
    const wrapper = shallow(<Accordion header="test header">test</Accordion>)
    expect(wrapper).toMatchSnapshot()
  })

  it('Allows us to set props', () => {
    const wrapper = mount(
      <Accordion header="test header" expanded>
        test
      </Accordion>
    )
    expect(wrapper.props().header).toEqual('test header')
    expect(wrapper.props().expanded).toEqual(true)
    wrapper.setProps({ header: 'foo' })
    wrapper.setProps({ expanded: false })
    expect(wrapper.props().header).toEqual('foo')
    expect(wrapper.props().expanded).toEqual(false)
  })

  it('setExpanded works', () => {
    const mockHandleChangeText = jest.fn()
    const component = mount(
      <Accordion
        header="test header"
        expanded
        setExpanded={mockHandleChangeText}
      >
        test
      </Accordion>
    )
    component
      ?.find(MuiAccordion)
      ?.props()
      ?.onChange?.(
        {
          target: { checked: true },
        } as React.ChangeEvent<HTMLInputElement>,
        true
      )
    expect(mockHandleChangeText).toHaveBeenCalledWith(false)
  })
})
