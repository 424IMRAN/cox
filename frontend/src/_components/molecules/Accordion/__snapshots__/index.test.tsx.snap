// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion  Component Renders correctly 1`] = `
<div
  className="makeStyles-root-1"
>
  <WithStyles(ForwardRef(Accordion))
    className="makeStyles-noBoxShadow-5"
    expanded={false}
    onChange={[Function]}
  >
    <WithStyles(ForwardRef(AccordionSummary))
      expandIcon={
        <Memo
          className="makeStyles-heading-2"
        />
      }
    >
      <Typography
        align="left"
        className="makeStyles-heading-2"
        variant="subtitle2"
      >
        test header
      </Typography>
    </WithStyles(ForwardRef(AccordionSummary))>
    <WithStyles(ForwardRef(AccordionDetails))>
      test
    </WithStyles(ForwardRef(AccordionDetails))>
  </WithStyles(ForwardRef(Accordion))>
</div>
`;
